docs/source/specs/typespec/main.tsp

```diff
@@ -0,0 +1,251 @@
+}
+
+interface CreateResource<BodyType> {
+    create<ModelType>(@body body: BodyType): CreatedResponseForModel<ModelType> | 
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorResponse400 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface ListResource<T> {
+    list(...ListQueryParams): T[] |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface ReadResource<T> {
+    read(
+        @doc("Unique identification")
+        @path uuid: UUID;
+    ): T |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorResponse404 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface UpdateResource<B, R, E> {
+    update(
+        @doc("Unique identification")
+        @path uuid: UUID;
+        @body body: B
+    ): OkResponseForModel<R> |
+    Errors.ErrorResponse400AlreadyExist<E> |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+interface DeleteResource {
+    delete(
+        @doc("Unique identification")
+        @path uuid: UUID;
+    ): NoContentResponseResourceDeleted |
+    Errors.ErrorResponse401 |
+    Errors.ErrorResponse403 |
+    Errors.ErrorBasicResponse<500>;
+}
+
+@route("/workspaces")
+@tag("Workspaces")
+@doc("Operations about workspaces")
+namespace Workspaces {
+    model WorkspaceForRead {
+        @doc("Workspace A")
+        name: string = "Workspace A";
+        @doc("Description of Workspace A")
+        description?: string = "Description of Workspace A";
+    }
+
+    model Workspace {
+        @key uuid: UUID;
+        ...WorkspaceForRead;
+        ...Timestamps;
+    }
+
+    @error
+    model ErrorMessage400WorkspaceAlreadyExist {
+        detail?: string = "Workspace already exist in tenant",
+        status?: string = "400",
+        source?: string = "workspace",
+    }
+
+    alias CreateResourceAlias = CreateResource<WorkspaceForRead>;
+    @doc("Create workspace in tenant")
+    @summary("Create workspace in tenant")
+    @post op create is CreateResourceAlias.create<Workspace>;
+
+    model WorkspaceListResponse is ListResponse<Workspace>;
+    alias ListResourceAlias = ListResource<WorkspaceListResponse>;
+    @doc("By default, responses are sorted in ascending order by workspace name")
+    @summary("List workspaces for tenant")
+    @get op list is ListResourceAlias.list;
+
+    alias ReadResourceAlias = ReadResource<Workspace>;
+    @doc("Get a workspace in tenant")
+    @summary("Get a workspace in tenant")
+    @get op read is ReadResourceAlias.read;
+
+    alias UpdateResourceAlias = UpdateResource<WorkspaceForRead, Workspace, ErrorMessage400WorkspaceAlreadyExist>;
+    @doc("Edit a workspace")
+    @summary("Edit the workspace name or description")
+    @put op update is UpdateResourceAlias.update;
+
+    @doc("Delete a workspace")
+    @summary("Delete the workspace")
+    @delete op delete is DeleteResource.delete;
```

Comment:
IMO there is a little bit too much templating / aliasing going on here. It's a little hard to understand and might hide errors.

For example, `ListResource.list` current returns type `T[]`, but is instantiated with a WorkspaceListResponse, which already includes an array of data. So it looks like an array of objects with metadata and an array of objects.

e.g. this is generated:

```
              schema:
                type: array
                items:
                  $ref: '#/components/schemas/Workspaces.WorkspaceListResponse'
```

I think you want

```
interface ListResource<T extends ListResponse> {
    list(...ListQueryParams): T |
    Errors.ErrorResponse401 |
    Errors.ErrorResponse403 |
    Errors.ErrorBasicResponse<500>;
}
```

It might be enough to just define reusable models for different request/response patterns, but drop the common interface operations. We could probably define an alias instead for some of the common error types to make those more reusable, for example.

```
alias CommonErrors = Errors.ErrorResponse401 | Errors.ErrorResponse403 | Errors.ErrorBasicResponse<500>;

// ...

  op create(@body body: CreateWorkspaceRequest): CreateWorkspaceResponse | CommonErrors | Problem.SomeUniqueProblemToWorkspaceCreation;
```

There is also some built-in stuff for REST interfaces in Typespec which might be interesting. Nothing is jumping out at me right away other than what you're already using basically but worth looking into I think. It has some interfaces which are meant to be extended I guess.

Reply:
Done 👍